Kudu.Services/FetchHelpers/OneDriveHelper.cs

```diff
@@ -0,0 +1,519 @@
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web.Script.Serialization;
+using Kudu.Contracts.Settings;
+using Kudu.Contracts.Tracing;
+using Kudu.Core;
+using Kudu.Core.Deployment;
+using Kudu.Core.Infrastructure;
+using Kudu.Core.Tracing;
+
+namespace Kudu.Services.FetchHelpers
+{
+    internal class OneDriveHelper
+    {
+        private const string JsonMediaType = "application/json";
+        private const string FormUrlEncodedMediaType = "application/x-www-form-urlencoded";
+        private const string UserAgentHeader = "User-Agent";
+        private const string MicrosoftAzure = "Microsoft Azure";
+        private const string CursorKey = "onedrive_cursor";
+        private const int DefaultBufferSize = 8192; // From http://msdn.microsoft.com/en-us/library/tyhc0kft.aspx
+        private const int MaxRetryCount = 3;
+        private const int MaxFilesPerSeconds = 5;
+
+        private readonly ITracer _tracer;
+        private readonly IDeploymentSettingsManager _settings;
+        private readonly IEnvironment _environment;
+
+        internal ILogger Logger
+        {
+            get;
+            set;
+        }
+
+        public OneDriveHelper(ITracer tracer,
+            IDeploymentSettingsManager settings,
+            IEnvironment environment)
+        {
+            _tracer = tracer;
+            _settings = settings;
+            _environment = environment;
+        }
+
+        internal async Task Sync(OneDriveInfo info)
+        {
+            string cursor = _settings.GetValue(CursorKey);
+            ChangesResult changes = null;
+            using (_tracer.Step("Getting delta changes with cursor: {0} ...", cursor))
+            {
+                changes = await GetChanges(info.AccessToken, info.RepositoryUrl, cursor);
+            }
+
+            if (changes == null || changes.Count == 0)
+            {
+                _tracer.Trace("No changes need to be applied.");
+                LogMessage(Resources.OneDriveNoChangesFound);
+                return;
+            }
+
+            using (_tracer.Step("Applying {0} changes ...", changes.Count))
+            {
+                LogMessage(Resources.OneDriveApplyingChanges, changes.Count);
+
+                // perform action seperately, so that can ensure timestamp on directory
+                // e.g two changes:
+                //  (new) file /a/b/c.txt
+                //  (new) dir  /a/b
+                //  if created dir first then create file. file creation will trigger folder timestamp change.
+                //  which will result in "/a/b" has timestamp from the monent of file creation instead of the timestamp value from server, where value supposed to be set by code specifically.
+                ApplyResult deletionResult = await ApplyChangesParallel(changes.DeletionChanges, info.AccessToken);
```

Comment:
See [HandlingDeletion](https://github.com/projectkudu/kudu/pull/1579/files#diff-3c0f9e394470bf81dbb6ba78d75f86d5R363)

dependency issue already handled, if folder deleted before file, FileNotFound or DirecotryNotFound will throw, and we will ignore these error.